Made backend choice a workflow option. Closes #567

cjllanwarne opened the pull request:
There's a few places here where things are a little bit ugly. That's half-deliberate - backend will become a per-call choice very soon and the ugliness will disappear. Guaranteed (TM)!

scottfrazer commented on engine/src/main/scala/cromwell/server/WorkflowManagerSystem.scala:
```diff
@@ -16,10 +17,11 @@ trait WorkflowManagerSystem {
   }
 
-  def backendType: String = ConfigFactory.load.getConfig("backend").getString("backend")
+  def backendType: List[String] = ConfigFactory.load.getConfig("backend").getStringList("backendsAllowed").toList
```
pluralize `backendType`?

Horneth commented on engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala:
```diff
@@ -15,24 +15,43 @@ import org.slf4j.LoggerFactory
 object CromwellBackend {
   private val log = LoggerFactory.getLogger(getClass.getName)
-  private var _backend: Option[Backend] = None
+  private var _backend: Option[Map[String, Backend]] = None
```
_backend**s** ?

cjllanwarne commented:
Could I request one review each from Broad (wheel chooses @scottfrazer) and Intel (@gauravs90?) please!

Thanks! :)

scottfrazer commented on engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala:
```diff
@@ -15,24 +15,43 @@ import org.slf4j.LoggerFactory
+//        _defaultBackend = Option(Backend.from(defaultBackend, actorSystem))
+//      case Some(x) if backendTypes.equals(x.keys.toList) => // TODO: This is only here because Main creates two instances of WorkflowManagerSystem. We should probably fix that.
+//      case Some(x) => throw new IllegalStateException(s"The set of backends has already been initialised to ${x.mkString}. Don't do that again (with ${backendTypes.mkString})!")
+//    }
```
Remove commented code

cjllanwarne replied:
Yeah, so in theory this check would be useful. In reality, initBackends is called so many times that this test causes problems. Hopefully in the future, we won't have to init backends at all :grinning:

scottfrazer commented on engine/src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala:
```diff
@@ -354,8 +355,10 @@ class WorkflowManagerActor(backend: Backend, config: Config)
 
     case class ActorAndStateData(actor: WorkflowActorRef, data: WorkflowManagerData)
 
+    // TODO: Obviously unpleasant, but when workflows don't have backends any more, this goes away. Hooray!
```
I don't know how everybody else thinks but I'm not a big fan of keeping comments like this in the code.

mcovarr replied:
I could live with it in the very short term if it was PBE rather than TODO

gauravs90 replied:
I second with @scottfrazer there. Just collected them all from intelliJ, there are some 40-odd TODOs (not necessarily all related to PBEs) currently overall, and it worries me how / when we get to clean those up.

mcovarr commented on engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala:
```diff
@@ -239,12 +239,36 @@ object WorkflowDescriptor {
   val OptionKeys: Set[String] = Set(WorkflowLogDirOptionKey, WorkflowOutputsOptionKey, CallLogsDirOptionKey)
 
+  // This is, frankly, an aberration. But don't worry - it's going away soon :D
+  // TODO: KILL IT WITH FIRE!
```
TODO --> PBE
FIRE! --> FIRE!!!

mcovarr commented on engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala:
```diff
@@ -239,12 +239,36 @@ object WorkflowDescriptor {
+  // This is, frankly, an aberration. But don't worry - it's going away soon :D
+  // TODO: KILL IT WITH FIRE!
+  def getBackendFromOptions(optionsString: String): Backend = {
+    val backendString = WorkflowOptions.fromJsonString(optionsString) flatMap { x => x.get("backend")}
```
`_.get("backend")`

mcovarr commented on engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala:
```diff
@@ -15,24 +15,43 @@ import org.slf4j.LoggerFactory
+  def backend(backendName: String): Backend = _backend match {
+    case Some(x: Map[String, Backend]) => x(backendName)
+    case None => throw new IllegalStateException("backend() called prior to initBackends")
+  }
```
`_backend map { _(backendName) } getOrElse throw new IllegalStateException(...)`

and I agree that variable name should be pluralized

mcovarr commented on engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala:
```diff
@@ -15,24 +15,43 @@ import org.slf4j.LoggerFactory
+    * to use a custom/mock/stub backend in a test.
+    *
+    */
+  def registerCustomBackend(backendName: String, backend: Backend) = {
```
If this is for tests can it have package visibility so it's not part of the public API?

cjllanwarne replied:
You mean protected? This isn't called from subclasses

geoffjentry replied:
i think he means something like `private[backend]`

gauravs90 replied:
Or maybe defined in the tests itself?

mcovarr commented on engine/src/main/scala/cromwell/engine/backend/CromwellBackend.scala:
```diff
@@ -15,24 +15,43 @@ import org.slf4j.LoggerFactory
   }
 
-  def backend() = _backend match {
+  def defaultBackend = _defaultBackend match {
```
`getOrElse`

mcovarr commented on engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala:
```diff
@@ -239,12 +239,36 @@ object WorkflowDescriptor {
+  def getBackendFromOptions(optionsString: String): Backend = {
+    val backendString = WorkflowOptions.fromJsonString(optionsString) flatMap { x => x.get("backend")}
+    backendString match {
+      case Success(x) => CromwellBackend.backend(x)
```
why not roll this into the `flatMap` above and follow that with `getOrElse` to return the default backend?

mcovarr commented on engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala:
```diff
@@ -239,12 +239,36 @@ object WorkflowDescriptor {
+  /**
+    * Warning: Uses the default backend for this workflow.
+    */
   def apply(id: WorkflowId, sourceFiles: WorkflowSourceFiles): WorkflowDescriptor = {
```
What *should* be using this `apply`?  Currently it's only called from tests

cjllanwarne replied:
Yes. There are a LOT of tests which are using this. It seemed slightly pointless to change them all to add Backends, only for the next change to have to go through and remove them again

geoffjentry replied:
@cjllanwarne is this what you were talking to me about the other day? If so I concur

cjllanwarne replied:
Yeah

gauravs90 replied:
Geez. This constructor had to be done with in the PR i am working on. I had to go through that pain :confused:

cjllanwarne replied:
Thanks @gauravs90! Yeah, I didn't *want* these constructors, they're no-longer current. But I didn't want to re-plumb everything else to work with my new temporary constructor which I'm pretty sure won't last very long either (apparently, it won't last past the end of your PR!)

mcovarr commented on engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala:
```diff
@@ -239,12 +239,36 @@ object WorkflowDescriptor {
+  /**
+    * Warning: Uses the default backend for this workflow.
+    */
   def apply(id: WorkflowId, sourceFiles: WorkflowSourceFiles, conf: Config): WorkflowDescriptor = {
```
same question for this `apply`, except it actually is being called from production code in one spot

cjllanwarne commented:
How do I make @coveralls stop commenting? We already have the status check...

geoffjentry commented:
@cjllanwarne I'd love to know. It's really annoying me today

gauravs90 commented on engine/src/main/scala/cromwell/engine/WorkflowDescriptor.scala:
```diff
@@ -239,12 +239,37 @@ object WorkflowDescriptor {
 
+  // PBE:This is, frankly, an aberration. But don't worry - it's going away soon :D
+  // KILL IT WITH FIRE!!!
+  def getBackendFromOptions(optionsString: String): Backend = {
```
I honestly think this should be defined elsewhere, (maybe in a separate file or in CromwellBackends, no opinions there, not sure if it can be for this PR though). It can then be extended in the subsequent PRs to initialize using reflection or returning a [potentially] list of backends that a WF may use (or anything else).

cjllanwarne replied:
I like this idea. I'll move it.

cjllanwarne commented:
@gauravs90 @francares Please note I've rebased and therefore had to update the ValidateActor to no longer require a backend on construction. I've also modified ValidateActorSpec to feed in the mock backend to the static CromwellBackend pool during testing.

scottfrazer commented:
reviewing now...

scottfrazer commented on engine/src/main/resources/application.conf:
```diff
@@ -70,7 +70,8 @@ call-caching {
   // Either "jes", "local", or "sge" (case insensitive)
-  backend = "local"
+  defaultBackend = "local"
+  backendsAllowed = [ "local" ]
```
So this seems to conflict with @mcovarr's email about how to specify the backends... I'm guessing this is meant to change soon? (is there a ticket?)

mcovarr replied:
Yes, this is an evolution.  The ticket for the bit I was looking at is #580.

scottfrazer replied:
thanks

scottfrazer commented on engine/src/test/scala/cromwell/RetryableCallsSpec.scala:
```diff
@@ -27,25 +25,27 @@ class RetryableCallsSpec extends CromwellTestkitSpec with Mockito {
+    """
+      |{
+      |  "backend": "customizedLocalBackend"
+      |}"""".stripMargin
```
I'm not crazy... there are definitely too many quotes, right?  My brain parser is reading this as

```
{
  "backend": "customizedLocalBackend"
}"
```

cjllanwarne replied:
Yeah... I wonder why the test is OK with this

scottfrazer replied:
Yeah I'm curious about that too... some parsers are actually okay with extraneous characters... but I feel like most are not.  Very odd.

cjllanwarne replied:
Oh yeah, I guess the JSON parser might be forgiving?

scottfrazer replied:
I just tried running a workflow with a trailing `"` in workflow options JSON file and it worked.  Interesting.

scottfrazer replied:
I actually had this exact same bug in Hermes for a while... it'd return valid parse trees even though it didn't consume the entire token stream.

scottfrazer commented on engine/src/test/scala/engine/workflow/ValidateActorSpec.scala:
```diff
@@ -21,7 +21,12 @@ class ValidateActorSpec extends TestKit(ActorSystem("ValidateActorSpecSystem"))
+    """
+      |{
+      |  "backend": "retryableCallsSpecBackend"
+      |}"""".stripMargin
```
too many quotes here too?

scottfrazer commented on engine/src/main/scala/cromwell/engine/workflow/ValidateActor.scala:
```diff
@@ -2,6 +2,9 @@ package cromwell.engine.workflow
 import akka.actor.{Actor, Props}
 import com.typesafe.scalalogging.LazyLogging
+import cromwell.engine.WorkflowDescriptor
+import cromwell.engine.WorkflowDescriptor
```
duplicate import

scottfrazer commented:
:+1:

gauravs90 commented on engine/src/test/scala/engine/workflow/ValidateActorSpec.scala:
```diff
@@ -59,22 +67,22 @@ class ValidateActorSpec extends TestKit(ActorSystem("ValidateActorSpecSystem"))
         }
       }
-      verify(backendMock, times(1)).assertWorkflowOptions(WorkflowOptions.fromJsonString(validOptionsJson).get)
+      verify(backendMock, times(1)).assertWorkflowOptions(WorkflowOptions.fromJsonString(customizedLocalBackendOptions).get)
```
Are these verify(..) statements are useful anymore, since it is using backendMock to verify the details (while the validation initializes a backend each time a `ValidateWorkflow` msg is received?

gauravs90 replied:
Never mind..Ignore this. I see you're passing in the mockBackend to the ValidateActor. All good :smile:

cjllanwarne replied:
I believe this line is saying "make sure that `assertWorkflowOptions` is called exactly once, with this argument" - which I believe is still a valid thing to check?

gauravs90 commented:
Small comment that needs to be verified. But other than that, :+1: 

[![Approved with PullApprove](https://img.shields.io/badge/pullapprove-approved-brightgreen.svg)](https://pullapprove.com/broadinstitute/cromwell/pull-request/586/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=broadinstitute/cromwell)
